tendermint/src/lite/public.rs

```diff
@@ -0,0 +1,170 @@
+}
+
+/// Attempt to update the store to the given untrusted height.
+/// This function is recursive: it uses a bisection algorithm
+/// to request data for intermediate heights as necessary.
+/// Ensures our last trusted header hasn't expired yet, and that
+/// data from the untrusted height can be verified, possibly using
+/// data from intermediate heights.
+/// This function is primarily for use by a light node.
+pub fn verify_and_update_bisection<TS, SH, C, L, R, S>(
+    untrusted_height: Height,
+    trust_threshold: &L,
+    trusting_period: &Duration,
+    now: &SystemTime,
+    req: &R,
+    store: &mut S,
+) -> Result<(), Error>
+where
+    TS: TrustedState<LastHeader = SH, ValidatorSet = C::ValidatorSet>,
+    SH: SignedHeader<Commit = C>,
+    C: Commit,
+    L: TrustThreshold,
+    R: Requester<SignedHeader = SH, ValidatorSet = C::ValidatorSet>,
+    S: Store<TrustedState = TS>,
+{
+    // Fetch the latest state and ensure it hasn't expired.
+    // Note when calling itself recursively, this check is redundant,
+    // but leaving it keeps things simple ...
+    let trusted_state = store.get(Height::from(0))?;
+    let trusted_sh = trusted_state.last_header();
+    is_within_trust_period(trusted_sh.header(), trusting_period, now)?;
+
+    // fetch the header and vals for the new height
+    let untrusted_sh = &req.signed_header(untrusted_height)?;
+    let untrusted_vals = &req.validator_set(untrusted_height)?;
+    let untrusted_next_vals = &req.validator_set(untrusted_height.increment())?;
+
+    // check if we can skip to this height and if it verifies.
+    match verify_single(
+        trusted_state,
+        untrusted_sh,
+        untrusted_vals,
+        untrusted_next_vals,
+        trust_threshold,
+    ) {
+        Ok(_) => {
+            // Successfully verified!
+            // Trust the new state and return.
+            let new_trusted_state = TS::new(untrusted_sh, untrusted_next_vals);
+            return store.add(&new_trusted_state);
+        }
+        Err(e) => {
+            // If something went wrong, return the error.
+            if e != Error::InsufficientVotingPower {
+                return Err(e);
+            }
+
+            // Insufficient voting power to update.
+            // Engage bisection, below.
+        }
+    }
+
+    // Get the pivot height for bisection.
+    let pivot_height: u64 = (trusted_sh.header().height().value() + untrusted_height.value()) / 2;
+
+    // Recursive call to update to the pivot height.
+    // When this completes, we will either return an error or
+    // have updated the store to the pivot height.
+    verify_and_update_bisection(
+        Height::from(pivot_height),
+        trust_threshold,
+        trusting_period,
+        now,
+        req,
+        store,
+    )?;
+
+    // Recursive call to update to the original untrusted_height.
+    verify_and_update_bisection(
+        untrusted_height,
+        trust_threshold,
+        trusting_period,
+        now,
```

Comment:
Are we safe passing at this point in time now? It has passed quite some time since original call.

Reply:
I think the use of `now` in the recursion is redundant since we're not getting a new source of `now`, unless we want this API to include a clock or somethign that can provide multiple instances of now. Hence the comment above:

```
    // Fetch the latest state and ensure it hasn't expired.
    // Note when calling itself recursively, this check is redundant,
    // but leaving it keeps things simple ...
```

We could refactor a bit so that there's an inner bisection function that is called recursively that doesn't take the now. Happy to do that if it makes things clearer

Reply:
Addressed in 60c8b16059c21ecbd72319ded47dcb68b2428cb0 by breaking out an inner recursive function that doesn't take `now`. This should make it clear we only check `now` once.